### PR TITLE
Use Webview to manage cluster settings

### LINKED ISSUE #88: Use Webview to manage cluster settings
In vscode-kafka, we need to manage UI input forms for : 

 * cluster configuration (today we can create the cluster configuration with a wizard, but we cannot edit it,an UI forms could be a good solution) like vscode-sqltools provides for connection:

image

 * start consumer should open UI input forms to fill some format (the key type, etc) and parameters (offset) like Conduktor provides:
 
image

To manage that we must use Webview API because vscode doesn't provide some UI forms toolkit (perhaps in the future? see https://github.com/microsoft/vscode/issues/94741).

UI input forms requires databinding between UI and TS model and validation for TS model. 

There are several solutions:

 * use HTML + JS (without framework),my fear is that code will become awful.
 * use angular. There are several articles about that https://medium.com/@GrandSchtroumpf/vscode-extension-inside-a-nx-workspace-f9ad232a26e5 but I don't see a lot of vscode extension based onangular.
 * use reactjs : vscode-sqltools does that. There are several articles about this topic like https://medium.com/younited-tech-blog/reactception-extending-vs-code-extension-with-webviews-and-react-12be2a5898fd
 * use vscode-page (based on handlerbar). https://dev.to/foxgem/simplifying-vs-code-webview-development-with-vscode-page-13c3 It's a nice article but the project have not a lot of stars
 * other?

@jlandersen @fbricon have you some opinions about that?

### PR DESCRIPTION
**Pull request by angelozerr**
Use Webview to manage cluster settings

Fixes #88

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by fbricon** [comment]
Screenshot 2021-05-26 at 12 05 54
- Page title should be "Edit Kafka cluster" or (New Kafka cluster)
- Header on top should be removed
- Cluster block should move up
- labels and text fields should be aligned and having the same width
- Authentication "None" should have a label
- Hide Back/Next, Finish -> Save

**Comment by angelozerr** [comment]
Please note now I'm consuming 0.2.5 vscode-wizard which has been published (which fixes some bugs that you mentionned)

> Page title should be "Edit Kafka cluster" or (New Kafka cluster)

Fixed (for the moment I manage only Edit)

> Header on top should be removed

Fixed by 0.2.5

> Cluster block should move up

Sorry I don't understand what you mean?

> labels and text fields should be aligned and having the same width

Fixed, but there is a problem with select that I have reported in https://github.com/redhat-developer/vscode-wizard/issues/33

> Authentication "None" should have a label

Fixed

> Hide Back/Next, Finish -> Save

TODO

**Comment by angelozerr** [comment]
I pushed a new commit, I have started to declare the UI for SSL:

image

Please note:

 * we should manage a browse button to select the proper files. See https://github.com/redhat-developer/vscode-wizard/issues/34
 * for the moment it's the UI, I need to write code to use values of this fields.

**Comment by rgrunber** [comment]
I'd probably use the full name of the term.
Ca -> Certificate Authority (CA)
Cert -> Certificate

I don't think the SSL checkbox should exist. The underlying field, `cluster.ssl` should simply be set to true when the other 3 fields are valid/non-empty.

**Comment by fbricon** [comment]
@rgrunber no you can set SSL=true without specifying certs

**Comment by angelozerr** [comment]
>I'd probably use the full name of the term.
>Ca -> Certificate Authority (CA)
>Cert -> Certificate

Fixed

>I don't think the SSL checkbox should exist. The underlying field, cluster.ssl should simply be set to true when the other 3 fields are valid/non-empty.

As @fbricon said we can enable SSL without certificates

**Comment by angelozerr** [comment]
Just for your information I created a PR https://github.com/redhat-developer/vscode-wizard/pull/37 to improve UI, here the result:

image

**Comment by angelozerr** [comment]
The PR is ready now. The only thing that I should add is the passphrase field (I think), but I'm wiating for feedback from https://github.com/jlandersen/vscode-kafka/issues/191#issuecomment-871999125

**Comment by fbricon** [comment]
need to document that you need to run "npm run compile" at least once before debugging the extension in vscode

**Comment by fbricon** [comment]
Screenshot 2021-07-01 at 15 37 34

- New Cluster(s) title should be "Add New Cluster(s)"
- cluster provider: remove the "none" option, have "configure manually"  preselected
- cluster provider combo should not crop selected provider

If vscode-rhoas is installed, you select the Red Hat provider, click finish, the wizard stays opened. It should close

**Comment by fbricon** [comment]
More to fix:

Screenshot 2021-07-01 at 15 46 08

**Comment by fbricon** [comment]
Input text alignment is borked in case of an error
Screenshot 2021-07-01 at 15 56 14

**Comment by fbricon** [comment]
When selecting SASL/*, the SSL box is not automatically checked, rendering the cluster unable to connect.

**Comment by fbricon** [comment]
When creating a new cluster, the wizard is stil opened after saving. If you click save again, a "Cluster name 'foo' already exists." error is shown. The wizard should be closed, or at the very least should enter edit mode.

**Comment by fbricon** [comment]
When editing a new cluster:
- change name
- save
- check ssl
- save => cluster name already exists

**Comment by fbricon** [comment]
In the SSL section, rename:
Key => Client key
Certificate => Client certificate

**Comment by fbricon** [comment]
certificate/key files should allow ".pem" and other file extensions (see https://kafka.js.org/docs/configuration#ssl and https://www.ssl.com/guide/pem-der-crt-and-cer-x-509-encodings-and-conversions/)

**Comment by angelozerr** [comment]
> New Cluster(s) title should be "Add New Cluster(s)"

fixed

> cluster provider: remove the "none" option, have "configure manually" preselected

fixed

> cluster provider combo should not crop selected provider

it's an issue for vscode-wizard, see https://github.com/redhat-developer/vscode-wizard/issues/58

> If vscode-rhoas is installed, you select the Red Hat provider, click finish, the wizard stays opened. It should close. If you click finish a 2nd time, it closes.

That's strange, I cannot reproduce on Windows OS?

**Comment by fbricon** [comment]
When renaming a cluster, the selected cluster in the status bar is not renamed

**Comment by angelozerr** [comment]
> When selecting SASL/*, the SSL box is not automatically checked, rendering the cluster unable to connect.

I cannot do that with vscode-wizard, but I validate that:

image

> When creating a new cluster, the wizard is stil opened after saving

I cannot reproduce it -(

> If you click save again, after checking the ssl box, a "Cluster name 'foo' already exists." error is shown. The wizard should be closed, or at the very least should enter edit mode.

fixed

> When editing an existing cluster:
>
>change name
>save
>check SSL => cluster name already exists

Same error than below, fixed

> certificate/key files should allow ".pem" and other file extensions (see https://kafka.js.org/docs/configuration#ssl and https://www.ssl.com/guide/pem-der-crt-and-cer-x-509-encodings-and-conversions/)

I add other extensions according to https://www.tutorialsteacher.com/https/ssl-certificate-format 

Here the filters now:

image

I allow *.* because I fear to forget some file extensions.

**Comment by angelozerr** [comment]
> When renaming a cluster, the selected cluster in the status bar is not renamed

fixed

**Comment by angelozerr** [comment]
@fbricon  everything should be fixed exept:

 * UI problem, I have created an issue in vscode-wizard https://github.com/redhat-developer/vscode-wizard/issues/58
 * disabled of  fileds according to combo mechanism, select SSL checkbox when mechanims != auto, I hav ecreated an issue in vscode-wizard https://github.com/redhat-developer/vscode-wizard/issues/59

File: src/wizards/validators.ts [hunk 1]
```diff
@@ -1,27 +1,33 @@
...
     return CommonsValidator.validateFieldUniqueValue(CLUSTER_FIELD, cluster, existingClusterNames);
 }
 
-export async function validateAuthentificationUserName(userName: string): Promise<string | undefined> {
+export function validateAuthentificationUserName(userName: string): string | undefined {
     return CommonsValidator.validateFieldRequired(USERNAME_FIELD, userName);
 }
 
+export function validateFile(filePath: string) : string | undefined {
+    if (!fs.existsSync(filePath)) {
```
Comment (fbricon) [inline src/wizards/validators.ts:26]: Should check it's an actual file

File: src/wizards/validators.ts [hunk 2]
```diff
@@ -1,27 +1,36 @@
...
+        return `The file '${filePath}' doesn't exists.`;
+    }
+    if (fs.lstatSync(filePath).isDirectory()) {
+        return `The '${filePath}' is a directory, please select a file.`;
```
Comment (fbricon) [inline src/wizards/validators.ts:30]: '${filePath}' is a directory, please select a file.

File: src/wizards/validators.ts [hunk 3]
```diff
@@ -1,27 +1,36 @@
...
 }
 
-export async function validateAuthentificationUserName(userName: string): Promise<string | undefined> {
+export function validateAuthentificationUserName(userName: string): string | undefined {
     return CommonsValidator.validateFieldRequired(USERNAME_FIELD, userName);
 }
 
+export function validateFile(filePath: string) : string | undefined {
+    if (!fs.existsSync(filePath)) {
+        return `The file '${filePath}' doesn't exists.`;
```
Comment (fbricon) [inline src/wizards/validators.ts:27]: doesn't exist.

File: src/wizards/validators.ts (hunk 2 above)
Comment (angelozerr) [reply src/wizards/validators.ts]: fixed

File: src/wizards/validators.ts (hunk 3 above)
Comment (angelozerr) [reply src/wizards/validators.ts]: fixed

File: src/wizards/validators.ts (hunk 1 above)
Comment (angelozerr) [reply src/wizards/validators.ts:26]: fixed

**Comment by fbricon** [comment]
Nice work @angelozerr! Thanks!

**Pull request merged**